### PR TITLE
✨ Add Validation webhooks to cluster api manager

### LINKED ISSUE #1757: Validation webhooks
**User Story**

As a user, I would like to get early validation errors if the supplied configuration is invalid.

As a Cluster API developer, I would like to be able to add strict validation to Cluster API types.

**Detailed Description**

[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature
/milestone v0.3.0

### PR DESCRIPTION
**Pull request by wfernandes**
Performs validation for Machines, MachineSets and MachineDeployments.
For Machines it validates that Bootstrap has either ConfigRef or Data
set.
For MachineSets and MachineDeployments it verifies selectors match
template labels.

Signed-off-by: Warren Fernandes <wfernandes@pivotal.io>

**What this PR does / why we need it**:
Performs validation for Machines, MachineSets and MachineDeployments.
For Machines it validates that Bootstrap has either ConfigRef or Data set.
For MachineSets and MachineDeployments it verifies selectors match template labels.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1757

### CODE REVIEW

**Comment by wfernandes** [comment]
/retest

Ran all tests locally without reproducing errors. 
`make test`
`./scripts/ci-test.sh`

Going to give it another go.

**Comment by wfernandes** [comment]
@detiber When you get the chance could you please review this PR. I noticed that you recently added validation to KubeadmControlPlane so the context is very similar.

I also fixed the service 
https://github.com/kubernetes-sigs/cluster-api/blob/08b29593e201a02a9805b8f839e5e22b789b0ff3/config/webhook/service.yaml#L12
to have `cluster-api-controller-manager` as the selector to match
https://github.com/kubernetes-sigs/cluster-api/blob/08b29593e201a02a9805b8f839e5e22b789b0ff3/config/default/manager_label_patch.yaml#L14

**Comment by ncdc** [comment]
FYI, I'm reviewing this now

File: controllers/machineset_controller.go [hunk 1]
```diff
@@ -161,26 +161,19 @@ func (r *MachineSetReconciler) reconcile(ctx context.Context, machineSet *cluste
...
-	// TODO(vincepri): Move to a validation (admission) webhook when supported.
 	selector, err := metav1.LabelSelectorAsSelector(&machineSet.Spec.Selector)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrapf(err, "failed to parse MachineSet %q label selector", machineSet.Name)
 	}
-
-	if !selector.Matches(labels.Set(machineSet.Spec.Template.Labels)) {
-		return ctrl.Result{}, errors.Errorf("failed validation on MachineSet %q label selector, cannot match any machines ", machineSet.Name)
-	}
+	// Copy label selector to its status counterpart in string format.
```
Comment (ncdc) [inline controllers/machineset_controller.go]: nit: this didn't need to move, did it?

File: api/v1alpha3/machine_webhook.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func (m *Machine) validateMachine() error {
+	if m.Spec.Bootstrap.ConfigRef == nil && m.Spec.Bootstrap.Data == nil {
+		return errors.Errorf(
+			"Expected at least one of `Bootstrap.ConfigRef` or `Bootstrap.Data` to be populated for Machine %q in namespace %q",
+			m.Name, m.Namespace,
+		)
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (wfernandes) [reply api/v1alpha3/machine_webhook.go]: Sure. I was actually hoping that someone would bring this up because I noticed the documentation and wasn't sure what the "correct" form was.

File: config/crd/kustomizeconfig.yaml [hunk 3]
```diff
@@ -7,11 +7,9 @@ nameReference:
     group: apiextensions.k8s.io
     path: spec/conversion/webhookClientConfig/service/name
 
-namespace:
```
[the remark this answers is not in the document]
Comment (wfernandes) [reply config/crd/kustomizeconfig.yaml]: Hmm. This must be an update. I was looking at the cronjob example.
https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/cronjob-tutorial/testdata/project/config/crd/kustomizeconfig.yaml
I'm fine with removing this change.

File: controllers/machineset_controller.go (hunk 1 above)
Comment (wfernandes) [reply controllers/machineset_controller.go]: It didn't. When I was refactoring I thought it would make sense to move the only usage of `selector` soon after it was created.

Comment (ncdc) [reply controllers/machineset_controller.go]: That's fair. When I review, one thing I look out for is keeping the change set to a minimum, so I ask questions like this one. Thank you for pointing out your reasoning. It's fine to move it so it's closer. Thanks.

File: api/v1alpha3/machine_webhook_test.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestMachineValidation(t *testing.T) {
+	t.Run("should validate machine bootstrap", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (wfernandes) [reply api/v1alpha3/machine_webhook_test.go]: I thought of keeping that to provide a description of what is being tested specifically. 
In the future if more validation needs to be added, testing that could be within another `t.Run("should validate something else"...`and a cleaner division would be defined.

[the remark this answers is not in the document]
Comment (ncdc) [reply api/v1alpha3/machine_webhook_test.go]: I would do separate test functions for that. For gingko-based tests, that's where you typically use `Describe` or `Context`, but I've never seen it used with Go's `testing`-based tests.

File: api/v1alpha3/machine_webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		{
+			name:      "should not return error if config ref is set",
+			bootstrap: Bootstrap{ConfigRef: &corev1.ObjectReference{}, Data: nil},
+			expectErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
```
Comment (ncdc) [inline api/v1alpha3/machine_webhook_test.go]: nit: because we aren't calling naked `Expect()` methods any more, I would make this `gomega.NewWithT` and change the gomega import so it's no longer a dot (`.`) import.

File: api/v1alpha3/machinedeployment_webhook.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			allErrs,
+			field.Invalid(field.NewPath("spec", "selector"), m.Spec.Selector, err.Error()),
+		)
+	} else if !selector.Matches(labels.Set(m.Spec.Template.Labels)) {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "template", "labels"),
+				m.Spec.Template.Labels,
+				"failed validation on label selector, cannot match Machine template labels",
```
Comment (ncdc) [inline api/v1alpha3/machinedeployment_webhook.go]: This prints out as, e.g.:

```
spec.template.labels: Invalid value: "a=b": failed validation on label selector, cannot match Machine template labels
```

If you change the detail message to `fmt.Sprintf("must match spec.selector %q", selector.String()),`, the output looks like this:

```
spec.template.labels: Invalid value: "a=b": must match spec.selector "c=d"
```

WDYT?

File: api/v1alpha3/machinedeployment_webhook_test.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+			name:      "should return error for invalid selector",
+			selectors: map[string]string{"-123-foo": "bar"},
+			labels:    map[string]string{"-123-foo": "bar"},
+			expectErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
```
Comment (ncdc) [inline api/v1alpha3/machinedeployment_webhook_test.go]: nit: same comment about not using the dot import

File: api/v1alpha3/machineset_webhook.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	return nil
+}
+
+func (m *MachineSet) validate() error {
+	var allErrs field.ErrorList
+	selector, err := metav1.LabelSelectorAsSelector(&m.Spec.Selector)
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "selector"), m.Spec.Selector, "failed to parse label selector"),
```
Comment (ncdc) [inline api/v1alpha3/machineset_webhook.go]:
```suggestion
			field.Invalid(field.NewPath("spec", "selector"), m.Spec.Selector, err.Error()),
```

File: api/v1alpha3/machineset_webhook.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+		)
+	} else if !selector.Matches(labels.Set(m.Spec.Template.Labels)) {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "template", "labels"),
+				m.Spec.Template.Labels,
+				"failed validation on label selector, cannot match Machine template labels",
```
Comment (ncdc) [inline api/v1alpha3/machineset_webhook.go]: Same comment/question as for machine deployments

File: api/v1alpha3/machineset_webhook_test.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+			name:      "should return error for invalid selector",
+			selectors: map[string]string{"-123-foo": "bar"},
+			labels:    map[string]string{"-123-foo": "bar"},
+			expectErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
```
Comment (ncdc) [inline api/v1alpha3/machineset_webhook_test.go]: nit: dot import comment 😄

File: config/default/kustomization.yaml [hunk 11]
```diff
@@ -10,11 +8,7 @@ namespace: capi-system
 # field above.
 namePrefix: capi-
 
-# Labels to add to all resources and selectors.
```
Comment (ncdc) [inline config/default/kustomization.yaml]: I would retain this, or if we want to remove it, do it in a separate PR. WDYT?

Comment (wfernandes) [reply config/default/kustomization.yaml]: Removing it in a separate PR is fine with me.

File: api/v1alpha3/machinedeployment_webhook.go (hunk 6 above)
Comment (wfernandes) [reply api/v1alpha3/machinedeployment_webhook.go]: Yeah, I just tested that out and it is definitely an improvement! It's still verbose but gets the point through.

This is the output I got.

> Error from server (MachineDeployment.cluster.x-k8s.io "test1-md-0" is invalid: spec.template.labels: Invalid value: map[string]string{"nodepool":"nodepool3"}: must match spec.selector "nodepool=nodepool-0"): error when creating "/tmp/md.yaml": admission webhook "validation.machinedeployment.cluster.x-k8s.io" denied the request: MachineDeployment.cluster.x-k8s.io "test1-md-0" is invalid: spec.template.labels: Invalid value: map[string]string{"nodepool":"nodepool3"}: must match spec.selector "nodepool=nodepool-0"

File: config/default/kustomization.yaml [hunk 12]
```diff
@@ -10,11 +8,11 @@ namespace: capi-system
...
-#commonLabels:
+# Labels to add to all resources and selectors.<Paste>
```
Comment (ncdc) [inline config/default/kustomization.yaml]: Bad undo? 😄

File: config/default/kustomization.yaml [hunk 13]
```diff
@@ -10,11 +8,11 @@ namespace: capi-system
...
+#commonLabels
```
Comment (ncdc) [inline config/default/kustomization.yaml]: The original in this has a `:` at the end

File: config/default/kustomization.yaml (hunk 12 above)
Comment (wfernandes) [reply config/default/kustomization.yaml]: woops! my bad!

File: config/crd/kustomization.yaml [hunk 14]
```diff
@@ -1,5 +1,3 @@
-apiVersion: kustomize.config.k8s.io/v1beta1
-kind: Kustomization
```
Comment (detiber) [inline config/crd/kustomization.yaml]: I'm not sure we should remove the type info here

File: config/default/kustomization.yaml [hunk 15]
```diff
@@ -1,5 +1,3 @@
-apiVersion: kustomize.config.k8s.io/v1beta1
-kind: Kustomization
```
Comment (detiber) [inline config/default/kustomization.yaml]: I'm not sure we should include these changes...  With more recent versions of kustomize bases is deprecated in favor of resources.

Comment (ncdc) [reply config/default/kustomization.yaml]: This is consistent with new kubebuilder (2.1.x) generated yaml

Comment (ncdc) [reply config/default/kustomization.yaml]: Dropping the apiVersion and kind, that is.

Comment (detiber) [reply config/default/kustomization.yaml]: Right, but it isn't necessarily consistent with the latest kustomize docs. I'm wondering if we should work on getting kubebuilder updated for this instead.

Comment (wfernandes) [reply config/default/kustomization.yaml]: I removed these because I was using cluster-api-provider-aws as a reference which didn't include them. Maybe we can create an issue on kubebuilder to track this.

**Comment by wfernandes** [comment]
@ncdc @detiber Are there any blockers on this that y'all would like me to address? I was hoping to build on top of this work to make a PR for https://github.com/kubernetes-sigs/cluster-api/issues/1794 😃

**Comment by detiber** [comment]
No blockers from me
/approve

File: config/default/kustomization.yaml [hunk 16]
```diff
@@ -14,7 +12,7 @@ namePrefix: capi-
 #commonLabels:
 #  someName: someValue
 
-resources:
```
Comment (ncdc) [inline config/default/kustomization.yaml]: @detiber I assume we should undo this change, because `bases` was deprecated in kustomize v2.1.0. Right?

**Comment by ncdc** [comment]
Just the 1 question about resources vs bases

File: config/default/kustomization.yaml (hunk 16 above)
Comment (detiber) [reply config/default/kustomization.yaml]: If we are trying to align with kubebuilder, then we should probably keep it as `bases`, it *should* still work for the time being...

Comment (ncdc) [reply config/default/kustomization.yaml]: Ok, works for me. We can always change it back 😄
